src/main-oe/sgxlkl_run_oe.c

```diff
@@ -2032,25 +2050,35 @@ int main(int argc, char* argv[], char* envp[])
         pthread_setname_np(sgxlkl_threads[i], "ENCLAVE");
     }
 
-    long exit_status = 0;
-
-    for (int i = 0; i < econf->ethreads; i++)
-    {
-        pthread_join(sgxlkl_threads[i], &return_value);
-        if (i == 0)
-        {
-            exit_status = (long)return_value;
-        }
-    }
-    sgxlkl_host_verbose_raw("\n");
-
-    if (oe_enclave)
+    // Wait for the terminating ethread to exit the enclave
+    pthread_mutex_lock(&terminating_ethread_exited_mtx);
+    ret = pthread_cond_wait(
+        &terminating_ethread_exited_cv, &terminating_ethread_exited_mtx);
+    if (ret != 0)
+        sgxlkl_host_fail("Failed to wait for enclave to finish: ret=%i\n", ret);
+    long exit_status = (long)enclave_return_status;
+    pthread_mutex_unlock(&terminating_ethread_exited_mtx);
+
+    // Only try to destroy enclave if all ethreads successfully exited
+    if (oe_enclave && exited_ethread_count == econf->ethreads)
```

Comment:
Why?  Don't we want to kill the enclave as long as one thread has delivered to us a return value?  The other threads are not doing anything we care about...

Reply:
Right, but the problem is that the OE terminate API call segfaults in these cases: https://github.com/lsds/sgx-lkl/pull/788#issue-467907462

Reply:
We probably need to file that as an OE bug and link to that issue here. If you can’t tear down an uncooperative enclave, that’s a problem. That said, what happens if we send a signal to all of the host pthreads that back our ethreads first? Does that not let us exit them all from the enclave and then tear it down?